qiskit/transpiler/passes/synthesis/plugin.py

```diff
@@ -0,0 +1,316 @@
+
+        If this returns ``True`` the plugin's ``run()`` method will be
+        passed a ``basis_gates`` kwarg with a list of gate names the target
+        backend supports. For example, ``['sx', 'x', 'cx', 'id', 'rz']``."""
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_coupling_map(self):
+        """Return whether the plugin supports taking ``coupling_map``
+
+        If this returns ``True`` the plugin's ``run()`` method will receive
+        two kwargs ``coupling_map`` and ``qubits``. The ``coupling_map`` kwarg
+        will be a :class:`~qiskit.transpiler.CouplingMap` object representing
+        the qubit connectivity of the target backend. The ``qubits`` kwarg will
+        receive a list of integers that represent the qubit indices in the
+        coupling map that unitary is on.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_approximation_degree(self):
+        """Return whether the plugin supports taking ``approximation_degree``
+
+        If this returns ``True`` the plugin's ``run()`` method will receive
+        a ``approximation_degree`` kwarg with a float value between 0 and 1
+        representing the closeness of the approximation to use (0: lowest,
+        1: highest).
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_natural_direction(self):
+        """Return whether the plugin supports a toggle for considering
+        directionality of 2-qubit gates as ``natural_direction``.
+
+        Refer to the documentation for :class:`~qiskit.transpiler.passes.UnitarySynthesis`
+        for the possible values and meaning of these values.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_pulse_optimize(self):
+        """Return whether the plugin supports a toggle to optimize pulses
+        during synthesis as ``pulse_optimize``.
+
+        Refer to the documentation for :class:`~qiskit.transpiler.passes.UnitarySynthesis`
+        for the possible values and meaning of these values.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_gate_lengths(self):
+        """Return whether the plugin supports taking ``gate_lengths``
+
+        ``gate_lengths`` will be a dictionary in the form of
+        ``{gate_name: {(qubit_1, qubit_2): length}}``. For example::
+
+            {
+            'sx': {(0,): 0.0006149355812506126, (1,): 0.0006149355812506126},
+            'cx': {(0, 1): 0.012012477900732316, (1, 0): 5.191111111111111e-07}
+            }
+
+        Do note that this dictionary might not be complete or could be empty
+        as it depends on the target backend reporting gate lengths on every
+        gate for each qubit.
+        """
+        pass
+
+    @property
+    @abc.abstractmethod
+    def supports_gate_errors(self):
+        """Return whether the plugin supports taking ``gate_errors``
```

Comment:
Here, also, I wish we specified a semantic meaning to the dictionary values. Are they bare floats recording average infidelities? Will we ever want anything else?

Reply:
This is just taken from the `BackendProperties` I was looking at the source to see if it's documented exactly what it is. All it says is "gate error estimates?: https://github.com/Qiskit/qiskit-terra/blob/6efe3c55015167788f1081e4531b6ff563328b94/qiskit/providers/models/backendproperties.py#L345 But I think your guess of it being bare floats for an average infidelities is probably it. Maybe @ajavadia can confirm?

If we do need something else (or more specific) we'll have to update the backend definition to enable the backend to report the extra information in a standardized way (the api supports it but it's free form extra/custom info the backend can report) and then we can add it to the plugin interface too.

Reply:
I have no extra functionality to propose — I myself have no need for diamond norm errors — but I do want to pin down what plug-in authors should expect as part of the contract.

(In a separate thread of questions, I learned that the backend commonly reports the same error rate for `x` and for `sx` — not because they suffer errors at the same rate, but because they are simultaneously used during 1Q randomized benchmarking, and no one bothers to separate the two effects. That's exactly the sort of thing I wish were tagged!)

Reply:
I expanded on the docstring here to say it's backend target specific in: https://github.com/Qiskit/qiskit-terra/pull/6124/commits/6f1eb05b5cf6b5c2ac83c9f32ca45ea506caea81